File: common-calculator/src/main/java/controller/CalculatorController.java
package controller;

import model.BasicCalculatorModel;
import model.AdvancedCalculatorModel;
import java.util.Scanner;

/**
 * 计算器控制器
 * 负责菜单显示、用户输入处理与流程控制，
 * 委托计算逻辑给 BasicCalculatorModel 和 AdvancedCalculatorModel
 */
public class CalculatorController {

    private final BasicCalculatorModel basicModel = new BasicCalculatorModel();
    private final AdvancedCalculatorModel advancedModel = new AdvancedCalculatorModel();
    private final Scanner scanner = new Scanner(System.in);

    /** 启动计算器主循环 */
    public void start() {
        while (true) {
            showMainMenu();
            int choice = getIntInput("请选择");
            switch (choice) {
                case 1 -> basicMenu();
                case 2 -> advancedMenu();
                case 0 -> {
                    System.out.println("感谢使用，再见！");
                    return;
                }
                default -> System.out.println("无效选项，请重新输入");
            }
        }
    }

    private void showMainMenu() {
        System.out.println("\n========== 计算器系统 ==========");
        System.out.println("如你所见，这是一个没有前端或客户端页面的极简控制台计算器项目");
        System.out.println("1. 基础的六则运算（加、减、乘、除、取余、倒数）");
        System.out.println("2. 一些高级点的运算（幂、阶乘、π、三角函数等）");
        System.out.println("0. 退出");
        System.out.println("=================================");
    }

    /**
     * 六则运算菜单
     * 处理加、减、乘、除、取余、倒数的输入与校验
     */
    private void basicMenu() {
        System.out.println("\n--- 六则运算 ---");
        System.out.println("1. 加法");
        System.out.println("2. 减法");
        System.out.println("3. 乘法");
        System.out.println("4. 除法");
        System.out.println("5. 取余");
        System.out.println("6. 倒数");
        System.out.println("0. 返回主菜单");

        int op = getIntInput("请输入运算符编号");
        if (op == 0) {
            return;
        }

        if (op < 1 || op > 6) {
            System.out.println("你好像选择了无效的运算选项");
            return;
        }

        double a = getDoubleInput("请输入第一个数");
        double b = 0;
        // 第六个选项，求倒数只需要取一个数
        if (op != 6) {
            b = getDoubleInput("请输入第二个数");
        }
        // 除法和取余等的零检查
        if (op == 4 && b == 0) {
            System.out.println("错误: 除数不能为0");
            return;
        }
        if (op == 5 && b == 0) {
            System.out.println("错误: 取余运算的模数不能为0");
            return;
        }
        if (op == 6 && a == 0) {
            System.out.println("错误: 0没有倒数");
            return;
        }

        try {
            double result = basicModel.calculate(a, b, op);
            System.out.printf("运算结果: %.3f%n", result);
        } catch (Exception e) {
            System.out.println("出现错误: " + e.getMessage());
        }
    }

    /**
     * 高级运算菜单
     * 处理幂、阶乘、π、三角函数、指数、对数等高级运算的输入与校验
     */
    private void advancedMenu() {
        System.out.println("\n--- 高级运算 ---");
        System.out.println("1. 幂运算 (a^b)");
        System.out.println("2. 阶乘 (n!)");
        System.out.println("3. 圆周率 π");
        System.out.println("4. sin(x)    （角度制）");
        System.out.println("5. cos(x)    （角度制）");
        System.out.println("6. tan(x)    （角度制）");
        System.out.println("7. e^x");
        System.out.println("8. ln(1+x)");
        System.out.println("0. 返回主菜单");

        int option = getIntInput("请选择");
        if (option == 0) {
            return;
        }

        if (option < 1 || option > 8) {
            System.out.println("你好像选择了无效的运算选项");
            return;
        }

        try {
            switch (option) {
                case 1 -> {
                    double base = getDoubleInput("底数");
                    int exp = getIntInput("指数");
                    System.out.printf("结果: %.3f%n", advancedModel.power(base, exp));
                }
                case 2 -> {
                    int n = getIntInput("整数 n");
                    System.out.printf("%d! = %.3f%n", n, (double) advancedModel.factorial(n));
                }
                case 3 -> System.out.printf("π = %.3f%n", advancedModel.getPi());
                case 4 -> {
                    double x = getDoubleInput("角度（度）");
                    System.out.printf("sin(%.3f°) ≈ %.3f%n", x, advancedModel.sinTaylor(x));
                }
                case 5 -> {
                    double x = getDoubleInput("角度（度）");
                    System.out.printf("cos(%.3f°) ≈ %.3f%n", x, advancedModel.cosTaylor(x));
                }
                case 6 -> {
                    double x = getDoubleInput("角度（度）");
                    System.out.printf("tan(%.3f°) ≈ %.3f%n", x, advancedModel.tanTaylor(x));
                }
                case 7 -> {
                    double x = getDoubleInput("x");
                    System.out.printf("e^%.3f ≈ %.3f%n", x, advancedModel.expTaylor(x));
                }
                case 8 -> {
                    double x = getDoubleInput("x (要求 x > -1)");
                    System.out.printf("ln(1+%.3f) ≈ %.3f%n", x, advancedModel.lnTaylor(x));
                }
                default -> System.out.println("你好像选择了无效的运算选项");
            }
        } catch (ArithmeticException e) {
            System.out.println("数学错误: " + e.getMessage());
        } catch (IllegalArgumentException e) {
            System.out.println("输入错误: " + e.getMessage());
        } catch (Exception e) {
            System.out.println("出现了未知错误: " + e.getMessage());
        }
    }

    /**
     * 读取整数输入（带校验循环）
     * @param prompt 输入提示文字
     * @return 用户输入的整数
     */
    private int getIntInput(String prompt) {
        System.out.print(prompt + ": ");
        while (true) {
            if (scanner.hasNextInt()) {
                return scanner.nextInt();
            } else {
                String invalid = scanner.next();
                System.out.print("请输入有效的整数（你输入了: " + invalid + "）: ");
            }
        }
    }

    /**
     * 读取浮点数输入（带校验循环）
     * @param prompt 输入提示文字
     * @return 用户输入的浮点数
     */
    private double getDoubleInput(String prompt) {
            System.out.print(prompt + ": ");
            while (true) {
                if (scanner.hasNextDouble()) {
                    return scanner.nextDouble();
                } else {
                    String invalid = scanner.next();
                    System.out.print("请输入有效的数字（你输入了: " + invalid + "）: ");
                }
            }
        }
}

File: common-calculator-c/controller/calculator_controller.c
#include "calculator_controller.h"
#include "../model/basic_calculator_model.h"
#include "../model/advanced_calculator_model.h"
#include <stdio.h>
#include <stdlib.h>

/**
 * 读取整数输入（带校验循环）
 */
static int get_int_input(const char *prompt) {
    int value;
    char buffer[100];
    while (1) {
        printf("%s: ", prompt);
        if (fgets(buffer, sizeof(buffer), stdin) == NULL) {
            exit(0);
        }
        if (sscanf(buffer, "%d", &value) == 1) {
            return value;
        }
    }
}

/**
 * 读取浮点数输入（带校验循环）
 */
static double get_double_input(const char *prompt) {
    double value;
    char buffer[100];
    while (1) {
        printf("%s: ", prompt);
        if (fgets(buffer, sizeof(buffer), stdin) == NULL) {
            exit(0);
        }
        if (sscanf(buffer, "%lf", &value) == 1) {
            return value;
        }
    }
}

/**
 * 六则运算菜单
 */
static void basic_menu(void) {
    printf("\n--- 六则运算 ---\n");
    printf("1. 加法\n");
    printf("2. 减法\n");
    printf("3. 乘法\n");
    printf("4. 除法\n");
    printf("5. 取余\n");
    printf("6. 倒数\n");
    printf("0. 返回主菜单\n");

    int op = get_int_input("请输入运算符编号");
    if (op == 0) {
        return;
    }

    if (op < 1 || op > 6) {
        printf("你好像选择了无效的运算选项\n");
        return;
    }

    double a = get_double_input("请输入第一个数");
    double b = 0;
    if (op != 6) {
        b = get_double_input("请输入第二个数");
    }

    if (op == 4 && b == 0) {
        printf("错误: 除数不能为0\n");
        return;
    }
    if (op == 5 && (int)b == 0) {
        printf("错误: 取余运算的模数不能为0\n");
        return;
    }
    if (op == 6 && a == 0) {
        printf("错误: 0没有倒数\n");
        return;
    }

    double result = basic_calculator_calculate(a, b, op);
    printf("运算结果: %.3f\n", result);
}

/**
 * 高级运算菜单
 */
static void advanced_menu(void) {
    printf("\n--- 高级运算 ---\n");
    printf("1. 幂运算 (a^b)\n");
    printf("2. 阶乘 (n!)\n");
    printf("3. 圆周率 \xcf\x80\n");
    printf("4. sin(x)    （角度制）\n");
    printf("5. cos(x)    （角度制）\n");
    printf("6. tan(x)    （角度制）\n");
    printf("7. e^x\n");
    printf("8. ln(1+x)\n");
    printf("0. 返回主菜单\n");

    int option = get_int_input("请选择");
    if (option == 0) {
        return;
    }

    if (option < 1 || option > 8) {
        printf("你好像选择了无效的运算选项\n");
        return;
    }

    switch (option) {
        case 1: {
            double base = get_double_input("底数");
            int exp = get_int_input("指数");
            printf("结果: %.3f\n", advanced_power(base, exp));
            break;
        }
        case 2: {
            int n = get_int_input("整数 n");
            long long fact = advanced_factorial(n);
            if (fact >= 0) {
                printf("%d! = %lld\n", n, fact);
            }
            break;
        }
        case 3:
            printf("\xcf\x80 = %.3f\n", advanced_get_pi());
            break;
        case 4: {
            double x = get_double_input("角度（度）");
            printf("sin(%.3f\xc2\xb0) \xE2\x89\x88 %.3f\n", x, advanced_sin_taylor(x));
            break;
        }
        case 5: {
            double x = get_double_input("角度（度）");
            printf("cos(%.3f\xc2\xb0) \xE2\x89\x88 %.3f\n", x, advanced_cos_taylor(x));
            break;
        }
        case 6: {
            double x = get_double_input("角度（度）");
            printf("tan(%.3f\xc2\xb0) \xE2\x89\x88 %.3f\n", x, advanced_tan_taylor(x));
            break;
        }
        case 7: {
            double x = get_double_input("x");
            printf("e^%.3f \xE2\x89\x88 %.3f\n", x, advanced_exp_taylor(x));
            break;
        }
        case 8: {
            double x = get_double_input("x (要求 x > -1)");
            printf("ln(1+%.3f) \xE2\x89\x88 %.3f\n", x, advanced_ln_taylor(x));
            break;
        }
        default:
            printf("你好像选择了无效的运算选项\n");
    }
}

/**
 * 显示主菜单
 */
static void show_main_menu(void) {
    printf("\n========== 计算器系统 ==========\n");
    printf("如你所见，这是一个没有前端或客户端页面的极简控制台计算器项目\n");
    printf("1. 基础的六则运算（加、减、乘、除、取余、倒数）\n");
    printf("2. 一些高级点的运算（幂、阶乘、\xcf\x80、三角函数等）\n");
    printf("0. 退出\n");
    printf("=================================\n");
}

void calculator_controller_start(void) {
    while (1) {
        show_main_menu();
        int choice = get_int_input("请选择");
        switch (choice) {
            case 1:
                basic_menu();
                break;
            case 2:
                advanced_menu();
                break;
            case 0:
                printf("感谢使用，再见！\n");
                return;
            default:
                printf("无效选项，请重新输入\n");
        }
    }
}
